test/test_common/test_time_system.h

```diff
@@ -41,5 +48,81 @@ class TestTimeSystem : public Event::TimeSystem {
+// There should only be one instance of any time-system resident in a test
+// process at once. This helper class is used with Test::Global to help enforce
+// that with an ASSERT. Each time-system derivation should have a helper
+// implementation which is referenced from a delegate (see
+// DelegatingTestTimeSystemBase). In each delegate, a SingletonTimeSystemHelper
+// should be instantiated via Test::Global<SingletonTimeSystemHelper>. Only one
+// instance of SingletonTimeSystemHelper per process, at a time. When all
+// references to the delegates are destructed, the singleton will be destroyed
+// as well, so each test-method will get a fresh start.
+class SingletonTimeSystemHelper {
+public:
+  SingletonTimeSystemHelper() : time_system_(nullptr) {}
+
+  void set(TestTimeSystem* time_system) {
+    ASSERT(time_system_ == nullptr);
+    time_system_.reset(time_system);
+  }
+
+  TestTimeSystem* timeSystem() { return time_system_.get(); }
+
+private:
+  std::unique_ptr<TestTimeSystem> time_system_;
+};
+
+// Implements the TestTimeSystem interface, delegating implementation of all
+// methods to a TestTimeSystem reference supplied by a timeSystem() method in a
+// subclass.
+template <class TimeSystemVariant> class DelegatingTestTimeSystemBase : public TestTimeSystem {
+public:
+  void sleep(const Duration& duration) override { timeSystem().sleep(duration); }
+
+  Thread::CondVar::WaitStatus
+  waitFor(Thread::MutexBasicLockable& mutex, Thread::CondVar& condvar,
+          const Duration& duration) noexcept EXCLUSIVE_LOCKS_REQUIRED(mutex) override {
+    return timeSystem().waitFor(mutex, condvar, duration);
+  }
+
+  SchedulerPtr createScheduler(Libevent::BasePtr& base_ptr) override {
+    return timeSystem().createScheduler(base_ptr);
+  }
+  SystemTime systemTime() override { return timeSystem().systemTime(); }
+  MonotonicTime monotonicTime() override { return timeSystem().monotonicTime(); }
+
+  TimeSystemVariant& operator*() { return timeSystem(); }
+
+  virtual TimeSystemVariant& timeSystem() PURE;
+};
+
+// Wraps a concrete time-system in a delegate that ensures thare is only one
+// time-system of any variant resident in a process at a time. Attempts to
+// instantiate multiple instances of the same type of time-system will simply
+// reference the same shared delegate, which will be deleted when the last one
+// goes out of scope. Attempts to instantiate different types of type-systems
+// will result in a RELEASE_ASSERT. See the testcases in
+// test_time_system_test.cc to understand the allowable sequences.
+template <class TimeSystemVariant>
+class DelegatingTestTimeSystem : public DelegatingTestTimeSystemBase<TimeSystemVariant> {
+public:
+  DelegatingTestTimeSystem() {
+    TestTimeSystem* time_system = singleton_->timeSystem();
+    if (time_system == nullptr) {
+      time_system_ = new TimeSystemVariant;
+      singleton_->set(time_system_);
+    } else {
+      time_system_ = dynamic_cast<TimeSystemVariant*>(time_system);
```

Comment:
Under what circumstances would `singleton_>timeSystem()` return something that was not `nullptr` but was also not dynamic castable to `TimeSystemVariant*`?

Reply:
Apologies, this was one of those comments that I made while working from the top down and failed to delete before flushing.